frontend/src/metabase/browse/components/ModelGroup.tsx

```diff
@@ -1,54 +1,199 @@
-import type {
-  Card,
-  CollectionEssentials,
-  SearchResult,
-} from "metabase-types/api";
+import { useDisclosure } from "@mantine/hooks";
+import { useEffect } from "react";
+import { c, msgid } from "ttag";
+import type { Card, CollectionId, SearchResult } from "metabase-types/api";
 import * as Urls from "metabase/lib/urls";
 
-import Link from "metabase/core/components/Link";
 import Search from "metabase/entities/search";
-import { useDispatch } from "metabase/lib/redux";
+import { useDispatch, useSelector } from "metabase/lib/redux";
 
-import { Box, Group, Icon, Text, Title } from "metabase/ui";
+import { Box, Icon, Title } from "metabase/ui";
 
+import { updateSetting } from "metabase/admin/settings/settings";
+import { isValidCollectionId } from "metabase/collections/utils";
 import { color } from "metabase/lib/colors";
+import { getSetting } from "metabase/selectors/settings";
 import { getCollectionName, sortModels } from "../utils";
 import {
+  CollectionCollapse,
+  CollectionExpandCollapseContainer,
+  CollectionHeader,
   CollectionHeaderContainer,
-  CollectionHeaderGroup,
   CollectionHeaderLink,
+  CollectionHeaderToggle,
+  CollectionSummary,
+  ContainerExpandCollapseButton,
+  FixedSizeIcon,
   ModelCard,
+  ModelCardLink,
   MultilineEllipsified,
 } from "./BrowseModels.styled";
-import { LastEdited } from "./LastEdited";
+
+type CollectionPrefs = Record<CollectionId, ModelVisibilityPrefs>;
+
+type ModelVisibilityPrefs = {
+  areAllModelsShown: boolean;
+  areSomeModelsShown: boolean;
+};
+
+const isRecordWithCollectionIdKeys = (
+  prefs: unknown,
+): prefs is Record<CollectionId, any> =>
+  typeof prefs === "object" &&
+  prefs !== null &&
+  Object.keys(prefs).every(key => isValidCollectionId(key));
+
+const isValidModelVisibilityPrefs = (
+  value: unknown,
+): value is ModelVisibilityPrefs =>
+  typeof value === "object" &&
+  value !== null &&
+  Object.keys(value).includes("areAllModelsShown") &&
+  Object.keys(value).includes("areSomeModelsShown") &&
+  Object.values(value).every(val => typeof val === "boolean");
+
+const isValidCollectionPrefs = (prefs: unknown): prefs is CollectionPrefs =>
+  isRecordWithCollectionIdKeys(prefs) &&
+  Object.values(prefs).every(val => isValidModelVisibilityPrefs(val));
+
+const tryToParseCollectionPrefs = (
+  json: string | null,
+): CollectionPrefs | null => {
+  try {
+    const parsed = JSON.parse(json || "");
+    return isValidCollectionPrefs(parsed) ? parsed : null;
+  } catch (e) {
+    if (e instanceof SyntaxError) {
+      return null;
+    }
+    throw e;
+  }
+};
```

Comment:
1) I don't think settings like this should be server-side, they're simple UI interactions, not data we need to persist. we don't want to cause a bunch of network requests and db write/reads every time the user simply manipulates the UI.
2) We should lift the expand/collapse show/hide state up since it's all one object and only pass down necessary properties - we have potentially thousands of collections here and we don't need them all doing the exact same parsing when we can do it once at the parent-component level.
3) these helper functions are great, but also perfect candidates to be extracted to a `utils` file to keep the component easier to understand at a glance.